src/essential/container.jl

```diff
@@ -9,36 +9,20 @@ function TracedModel(
     model::Model,
     sampler::AbstractSampler,
     varinfo::AbstractVarInfo,
-    rng::Random.AbstractRNG
-) 
+    rng::Random.AbstractRNG,
+)
     context = SamplingContext(rng, sampler, DefaultContext())
-    evaluator = _get_evaluator(model, varinfo, context)
-    return TracedModel{AbstractSampler,AbstractVarInfo,Model,Tuple}(model, sampler, varinfo, evaluator)
+    # FIXME: We're just dropping the `kwargs` here. I'm guessing this can
+    # cause issues if the model ends up mutating any of the keyword arguments.
+    args, _ = DynamicPPL.make_evaluate_args_and_kwargs(model, varinfo, context)
```

Comment:
Okay, so I've figured out why it breaks, but I think it's going to be difficult to fix it :confused: 

Libtask only handles nested `produce` in the case where each instruction it sees contains _at most_ one `produce`.

```julia
julia> using Libtask

julia> f(x) = (produce(x); produce(2x); produce(3x); return nothing)
f (generic function with 1 method)

julia> g(x) = f(x)
g (generic function with 1 method)

julia> task = Libtask.TapedTask(f, 1);

julia> consume(task), consume(task), consume(task)
(1, 2, 3)

julia> task = Libtask.TapedTask(g, 1);  # tracing of nested call

julia> consume(task)   # goes through all the `produce` calls before even calling the `callback` (which is `Libtask.producer`)
counter=1
tf=TapedFunction:
* .func => g
* .ir   =>
------------------
CodeInfo(
1 ─ %1 = Main.f(x)::Core.Const(nothing)
└──      return %1
)
------------------

ErrorException("There is a produced value which is not consumed.")Union{Ptr{Nothing}, Base.InterpreterIP}[Ptr{Nothing} @0x00007fa8d200eeeb, Ptr{Nothing} @0x00007fa8a0a30f29, Ptr{Nothing} @0x00007fa8a0a36844, Ptr{Nothing} @0x00007fa8a0a36865, Ptr{Nothing} @0x00007fa8e6b44f1d, Ptr{Nothing} @0x00007fa8a0a366e3, Ptr{Nothing} @0x00007fa8a0a36802, Ptr{Nothing} @0x00007fa8e6b44f1d, Ptr{Nothing} @0x00007fa8a0a35f25, Ptr{Nothing} @0x00007fa8a0a361dd, Ptr{Nothing} @0x00007fa8a0a36512, Ptr{Nothing} @0x00007fa8a0a3652f, Ptr{Nothing} @0x00007fa8e6b44f1d, Ptr{Nothing} @0x00007fa8e6b6656f]
ERROR: There is a produced value which is not consumed.
Stacktrace:
 [1] consume(ttask::TapedTask{typeof(g), Tuple{Int64}})
   @ Libtask ~/.julia/packages/Libtask/h7Kal/src/tapedtask.jl:153
 [2] top-level scope
   @ REPL[9]:1
```

This also makes me realize that Libtask.jl will also just silently do the wrong thing in certain cases where we use `@submodel` and the inner-model contains observations..

Reply:
And all of this effectively means that, with the current Libtask, we cannot support kwargs in a model with out converting it all to positional arguments..

Reply:
> In theory this can be fixed by adding submodels into lists that we need to trace into.

AFAIK there's no way to customize which functions to trace into.

Reply:
`TapedFunction` only traces the top-level, no more. It does not recurse into anything.

Reply:
Ah gotcha. But yeah, it should be possible. I'm actually trying to implement exactly that right now.

It should be possible to recursively call `Libtask._infer` (effectively, `Base.code_typed`) on the functions we want to recursively unroll.

Reply:
Btw, is there a reason why we didn't go for, say, IRTools? This recursively unrolls all the way I believe?

Reply:
I'd love to:) But this completely breaks SMC samplers for any model with kwargs :confused:

Reply:
So it will run with all kwarg models _but_ it will silently do the incorrect thing if the default kwargs are not the ones we're supposed to use. In this PR we currently just drop them completely, i.e. we end up using the default kwargs always. 

I don't think this quite qualifies as good enough:/